File: regional-monitor/frontend/src/api/keywordDna.ts
/**
 * 타지역키워드 DNA 파싱 솔루션 API 클라이언트.
 *
 * 백엔드:
 *   GET  /api/v1/keyword-dna/health
 *   GET  /api/v1/keyword-dna/dictionary/stats
 *   GET  /api/v1/keyword-dna/recommended
 *   POST /api/v1/keyword-dna/analyze
 *   POST /api/v1/keyword-dna/analyze/batch
 */
import { api } from './client'

export type DnaCategory = 'main' | 'action' | 'material' | 'place' | 'brand' | 'tag'

export interface DnaToken {
  token: string
  df: number
  weight: number
  share: number
}

export interface DnaGoldenCombo {
  combo: string
  main: string
  modifier: string
  modifier_category: DnaCategory
  weight: number
}

export interface DnaExample {
  name: string
  weight: number
  tokens: string[]
}

export interface DnaStats {
  matched: number
  total: number
  weight_matched: number
  total_weight: number
  elapsed_ms: number
}

export interface DnaResult {
  keyword: string
  normalized: string
  stats: DnaStats
  dna: Record<DnaCategory, DnaToken[]>
  golden: DnaGoldenCombo[]
  examples: DnaExample[]
  error?: string
}

export interface RecommendedItem {
  token: string
  category: DnaCategory
  df: number
  weight: number
}

export interface DictionaryStats {
  categories: DnaCategory[]
  stats: {
    business_count: number
    category_count: number
    total_weight: number
    token_count: number
    by_category: Record<DnaCategory, number>
  }
}

// ─── 비교 매트릭스 ───
export interface CompareRow {
  token: string
  category: DnaCategory
  kw_count: number
  total_weight: number
  weights: Record<string, number>
  dfs: Record<string, number>
  is_shared: boolean
  is_unique: boolean
}

export interface CompareSimilarity {
  kw1: string
  kw2: string
  jaccard: number
  cosine: number
  shared: string[]
  shared_count: number
}

export interface CompareSummary {
  keyword: string
  matched: number
  weight_matched: number
  share: number
  elapsed_ms: number
}

export interface CompareResult {
  keywords: string[]
  summary: CompareSummary[]
  matrix: CompareRow[]
  matrix_total: number
  similarity: CompareSimilarity[]
  shared_count: number
  unique_count: number
  error?: string
}

// ─── 네트워크 그래프 ───
export interface GraphNode {
  id: string
  category: DnaCategory
  weight: number
  df: number
  size: number
  is_center: boolean
}

export interface GraphEdge {
  source: string
  target: string
  weight: number
  df: number
}

export interface GraphResult {
  keyword: string
  normalized: string
  nodes: GraphNode[]
  edges: GraphEdge[]
  stats: { matched: number; node_count: number; edge_count: number; elapsed_ms: number }
  error?: string
}

// ─── 추천 ───
export type OppStatus = 'uncovered' | 'low_competition' | 'moderate' | 'saturated'

export interface OppCandidate {
  combo: string
  modifier: string
  modifier_category: DnaCategory
  market_weight: number
  market_df: number
  competition_weight: number
  competition_count: number
  opportunity: number
  status: OppStatus
  status_label: string
}

export interface RecommendResult {
  seed: string
  normalized: string
  candidates: OppCandidate[]
  stats: { seed_matched: number; candidate_count: number; elapsed_ms: number }
  error?: string
}

export const KeywordDnaApi = {
  health: () =>
    api.get<{ status: string; service: string; dictionary: any }>(
      '/api/v1/keyword-dna/health',
    ),
  dictionaryStats: () =>
    api.get<DictionaryStats>('/api/v1/keyword-dna/dictionary/stats'),
  recommended: (top = 50) =>
    api.get<{ count: number; items: RecommendedItem[] }>(
      `/api/v1/keyword-dna/recommended?top=${top}`,
    ),
  analyze: (
    keyword: string,
    opts: { top_per_category?: number; min_df?: number; examples?: number } = {},
  ) =>
    api.post<DnaResult>('/api/v1/keyword-dna/analyze', {
      keyword,
      top_per_category: opts.top_per_category ?? 15,
      min_df: opts.min_df ?? 2,
      examples: opts.examples ?? 30,
    }),
  analyzeBatch: (
    keywords: string[],
    opts: { top_per_category?: number; min_df?: number; examples?: number } = {},
  ) =>
    api.post<{ count: number; results: DnaResult[] }>(
      '/api/v1/keyword-dna/analyze/batch',
      {
        keywords,
        top_per_category: opts.top_per_category ?? 10,
        min_df: opts.min_df ?? 2,
        examples: opts.examples ?? 15,
      },
    ),
  compare: (
    keywords: string[],
    opts: { top_per_category?: number; min_df?: number } = {},
  ) =>
    api.post<CompareResult>('/api/v1/keyword-dna/compare', {
      keywords,
      top_per_category: opts.top_per_category ?? 12,
      min_df: opts.min_df ?? 2,
    }),
  graph: (
    keyword: string,
    opts: { max_nodes?: number; min_edge_weight?: number } = {},
  ) =>
    api.post<GraphResult>('/api/v1/keyword-dna/graph', {
      keyword,
      max_nodes: opts.max_nodes ?? 40,
      min_edge_weight: opts.min_edge_weight ?? 1.0,
    }),
  recommend: (
    seed: string,
    opts: { top?: number; min_modifier_df?: number } = {},
  ) =>
    api.post<RecommendResult>('/api/v1/keyword-dna/recommend', {
      seed,
      top: opts.top ?? 20,
      min_modifier_df: opts.min_modifier_df ?? 3,
    }),
}

File: regional-monitor/backend/app/api/keyword_dna.py
"""타지역키워드 DNA 파싱 솔루션 — API.

POST /api/v1/keyword-dna/analyze         — 단일 키워드 DNA 분석
POST /api/v1/keyword-dna/analyze/batch   — 다중 키워드 (최대 10개)
GET  /api/v1/keyword-dna/recommended     — 추천 키워드 (회선수 desc)
GET  /api/v1/keyword-dna/health
GET  /api/v1/keyword-dna/dictionary/stats
"""
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.api.deps import require_complete_profile
from app.services.keyword_dna import (
    analyze_keyword,
    list_known_keywords,
    load_dictionary,
    compare_keywords,
    build_graph,
    recommend_keywords,
)
from app.services.keyword_dna.dictionary import load_categories_list

router = APIRouter(prefix="/keyword-dna", tags=["keyword-dna"])


class AnalyzeRequest(BaseModel):
    keyword: str = Field(..., min_length=1, max_length=30)
    top_per_category: int = Field(default=15, ge=3, le=30)
    min_df: int = Field(default=2, ge=1, le=10)
    examples: int = Field(default=30, ge=5, le=100)


class AnalyzeBatchRequest(BaseModel):
    keywords: List[str] = Field(..., min_items=1, max_items=10)
    top_per_category: int = Field(default=10, ge=3, le=30)
    min_df: int = Field(default=2, ge=1, le=10)
    examples: int = Field(default=15, ge=5, le=50)


class CompareRequest(BaseModel):
    keywords: List[str] = Field(..., min_items=2, max_items=8)
    top_per_category: int = Field(default=12, ge=3, le=30)
    min_df: int = Field(default=2, ge=1, le=10)


class GraphRequest(BaseModel):
    keyword: str = Field(..., min_length=1, max_length=30)
    max_nodes: int = Field(default=40, ge=10, le=80)
    min_edge_weight: float = Field(default=1.0, ge=0.0)


class RecommendRequest(BaseModel):
    seed: str = Field(..., min_length=1, max_length=30)
    top: int = Field(default=20, ge=5, le=50)
    min_modifier_df: int = Field(default=3, ge=1, le=10)


@router.get("/health")
def health():
    d = load_dictionary()
    return {
        "status": "ok",
        "service": "keyword-dna",
        "dictionary": d.get("stats", {}),
    }


@router.get("/dictionary/stats")
def dictionary_stats(_user=Depends(require_complete_profile)):
    d = load_dictionary()
    return {
        "categories": d["categories"],
        "stats": d["stats"],
    }


@router.get("/recommended")
def recommended(
    top: int = 50,
    _user=Depends(require_complete_profile),
):
    return {
        "count": top,
        "items": list_known_keywords(top=top),
    }


@router.get("/categories")
def categories():
    """타지역업종리스트 (categories.xlsx) 카테고리·회선수 전체.

    공개 엔드포인트(인증 불필요) — '타지역 필수업종' 페이지에서 사용.
    """
    items = [{"category": c, "count": int(w)} for c, w in load_categories_list()]
    items.sort(key=lambda r: r["count"], reverse=True)
    total = sum(r["count"] for r in items)
    return {
        "count": len(items),
        "total_weight": total,
        "categories": items,
    }


@router.post("/analyze")
def analyze(
    req: AnalyzeRequest,
    _user=Depends(require_complete_profile),
):
    if not req.keyword.strip():
        raise HTTPException(status_code=422, detail="키워드가 비어 있습니다")
    try:
        return analyze_keyword(
            req.keyword.strip(),
            top_per_category=req.top_per_category,
            min_df=req.min_df,
            include_examples=req.examples,
        )
    except Exception as exc:  # pragma: no cover
        raise HTTPException(status_code=500, detail=f"분석 실패: {exc}") from exc


@router.post("/analyze/batch")
def analyze_batch(
    req: AnalyzeBatchRequest,
    _user=Depends(require_complete_profile),
):
    keywords = [k.strip() for k in req.keywords if k and k.strip()]
    if not keywords:
        raise HTTPException(status_code=422, detail="키워드 리스트가 비어 있습니다")
    results = []
    for kw in keywords:
        try:
            results.append(analyze_keyword(
                kw,
                top_per_category=req.top_per_category,
                min_df=req.min_df,
                include_examples=req.examples,
            ))
        except Exception as exc:  # pragma: no cover
            results.append({"keyword": kw, "error": str(exc)})
    return {"count": len(results), "results": results}


@router.post("/compare")
def compare(
    req: CompareRequest,
    _user=Depends(require_complete_profile),
):
    """다중 키워드 비교 매트릭스 — 토큰 × 키워드 가중치 + 유사도."""
    try:
        return compare_keywords(
            req.keywords,
            top_per_category=req.top_per_category,
            min_df=req.min_df,
        )
    except Exception as exc:  # pragma: no cover
        raise HTTPException(status_code=500, detail=f"비교 실패: {exc}") from exc


@router.post("/graph")
def graph(
    req: GraphRequest,
    _user=Depends(require_complete_profile),
):
    """단일 키워드 동시출현 네트워크 (nodes/edges)."""
    if not req.keyword.strip():
        raise HTTPException(status_code=422, detail="키워드가 비어 있습니다")
    try:
        return build_graph(
            req.keyword.strip(),
            max_nodes=req.max_nodes,
            min_edge_weight=req.min_edge_weight,
        )
    except Exception as exc:  # pragma: no cover
        raise HTTPException(status_code=500, detail=f"그래프 빌드 실패: {exc}") from exc


@router.post("/recommend")
def recommend(
    req: RecommendRequest,
    _user=Depends(require_complete_profile),
):
    """미커버/저경쟁 키워드 조합 추천 (opportunity score desc)."""
    if not req.seed.strip():
        raise HTTPException(status_code=422, detail="seed 키워드가 비어 있습니다")
    try:
        return recommend_keywords(
            req.seed.strip(),
            top=req.top,
            min_modifier_df=req.min_modifier_df,
        )
    except Exception as exc:  # pragma: no cover
        raise HTTPException(status_code=500, detail=f"추천 실패: {exc}") from exc
